kernel/posix/semaphore.c

```diff
@@ -136,3 +139,342 @@ int sem_wait(sem_t *semaphore)
 	k_sem_take(semaphore, K_FOREVER);
 	return 0;
 }
+
+
+/*
+ * Named semaphore design:
+ * - Opened table -> which semaphore is open by which thread.
+ * - Named semaphore pool -> global struct to allocate named sem.
+ */
+
+/*
+ * Used for error setting and exit
+ */
+#define SEM_ERRNO_SET(_code)      \
+	do {                      \
+		errno = (_code);  \
+		goto error;       \
+	} while (0)
+
+#define NSEM_SYS_N_MAX     (32) /* Limit of allocated nsem in system */
+#define NSEM_SYS_OPEN_MAX  (32) /* System limit of simultaneously open nsem */
+#define NSEM_PROC_OPEN_MAX  (4) /* Process limit of simultaneously open nsem */
+
+/*
+ * Private named semaphore
+ */
+struct _nsem {
+	sem_t sem;
+	const char *name;
```

Comment:
This won't work.  You need to manage storage for a copy of the name, you can't rely on the string pased to sem_open() being a literal and persisting beyond return.  This (weird) code is perfectly legal per POSIX but would fail in your setup:

    static char semname[] = "/foo";
    sem = sem_open(semname, O_CREAT);
    strcat(semname, "/bar");

With the implementation as it stands, the semaphore will not be finable via its original name anymore.

Reply:
I was not aware that it's valid to change the name of a semaphore. Mostly cause looks thread-unsafe to me.